storage/src/main/scala/coop/rchain/storage/Lmdb.scala

```diff
@@ -0,0 +1,1285 @@
+        throw new RChainException("Base Directory is invalid")
+      }
+    } else System.getProperty("user.dir") // not our problem if throws
+  val mapSizeIn: Long =
+    if (Lmdb.maxDbSize < maxSizeIn) {
+      Lmdb.maxDbSize
+    } else if (maxSizeIn < Lmdb.minDbSize) {
+      Lmdb.minDbSize
+    } else {
+      maxSizeIn
+    }
+
+  protected[storage] val dbDir = new File(baseDir.toString + "/" + dirName)
+  private var success = false
+  if (!dbDir.exists) {
+    success = dbDir.mkdir()
+    assert(success)
+  }
+
+  protected[storage] val dbDataFile = new File(dbDir.toPath + "/" + "data.mdb")
+  protected[storage] val dbLockFile = new File(dbDir.toPath + "/" + "lock.mdb")
+
+  val env: Env[ByteBuffer] =
+    if (isWritableIn)
+      create()
+        .setMaxDbs(1)
+        .setMapSize(mapSizeIn)
+        .open(dbDir, MDB_WRITEMAP)
+    else
+      create()
+        .setMaxDbs(1)
+        .setMapSize(mapSizeIn)
+        .open(dbDir)
+  val db: Dbi[ByteBuffer] =
+    if (isKeyToValuesIn) { env.openDbi(nameIn.get, MDB_CREATE, MDB_DUPSORT) } else {
+      env.openDbi(nameIn.get, MDB_CREATE)
+    }
+
+  // add a key-value row to the database
+  def put[K, V](key: K,
+                value: V,
+                txn: Option[Txn[ByteBuffer]] = None): Boolean = {
+    if (!isWritable) { return false }
+
+    if (key.isInstanceOf[String]
+        && getMaxKeySize < key.asInstanceOf[String].length) {
+      throw new RChainException("put(): key string is too long")
+    }
+    val bbKey: Option[ByteBuffer] = Bb.create(key)
+    val bbValue = Bb.create(value)
+    if (!bbKey.isDefined || !bbValue.isDefined) {
+      return false
+    }
+    putByteBuffer(bbKey.get, bbValue.get, txn)
+  }
+
+  def putByteBuffer(key: ByteBuffer,
+                    value: ByteBuffer,
+                    txnIn: Option[Txn[ByteBuffer]] = None): Boolean = {
+    if (!isWritable) { return false }
+
+    val txn =
+      if (txnIn == None) { env.txnWrite() } else { txnIn.get }
+    val cursor = db.openCursor(txn)
+    cursor.put(key, value)
+    cursor.close()
+    if (txnIn == None) { txn.commit() }
+    true
+  }
+
+  // The getXx[K](key:K) returns an array of Xs.
+  // This set of methods resist being expressed as a
+  // single generic method due to the lack of genericity
+  // of ByteBuffer.
+
+  def getInts[K](key: K,
+                 txnIn: Option[Txn[ByteBuffer]] = None): Option[Array[Int]] = {
+
+    if (key.isInstanceOf[Key]) {
+      return getInts(key.asInstanceOf[Key].term, txnIn)
+    }
+
+    if (!Lmdb.isStringOrPrimitive(key))
+      throw new RChainException("getInts(): key is not primitive or string")
+
+    var bbKey: Option[ByteBuffer] = None
+    if (Lmdb.isPrimitive(key)) {
+      bbKey = Bb.create(key)
+    } else {
+      bbKey = Some(Bb.strToBb(key.asInstanceOf[String]))
+    }
+    if (!bbKey.isDefined) {
+      throw new RChainException("getInts(): key fails translation to Bb")
+    }
+
```

Comment:
I would _definitely_ factor out these 16 initial lines into something like (I can't test this):
```Scala
  def makeBbKey[K](k: K): Option[ByteBuffer] =
    k match {
      case key: Key => makeBbKey(key.term)
      case term: Term =>
        if (Lmdb.isPrimitive(term)) {
          Bb.toBb(term)
        } else {
          Some(Bb.strToBb(term.asInstanceOf[String]))
        }
      case _ => None
    }
```